packages/frontend/src/hooks/useMulticallTokenBalance.tsx

```diff
@@ -1,86 +1,107 @@
+import { ERC20_ABI, ETH_CONTRACT_ADDRESS } from 'src/constants';
 import type { BalanceMap, MulticallToken } from 'src/types';
-import { formatTokenAmount } from "src/utils";
-import { useSwapFormValues } from 'src/hooks/useSwapFormValues';
+import { formatTokenAmount } from 'src/utils';
+import { useSifi } from 'src/providers/SDKProvider';
+import { TokenUsdPrice } from '@sifi/sdk';
+import Big from 'big.js';
 
 type UseMultiCallTokenBalance = {
   balanceMap: BalanceMap | null;
   refetch: () => void;
-}
+};
 
 const useMultiCallTokenBalance = (
   tokens: MulticallToken[],
   chainId: number
 ): UseMultiCallTokenBalance => {
-  const { fromChain } = useSwapFormValues();
+  const sifi = useSifi();
   const { address } = useAccount();
-  const [addressBalanceMap, setAddressBalanceMap] = useState<BalanceMap | null>(null)
+  const [addressBalanceMap, setAddressBalanceMap] = useState<BalanceMap | null>(null);
   const publicClient = usePublicClient({ chainId });
 
-  const balanceReadContracts = address?.startsWith('0x') ? tokens.map(token => ({
-    address: token.address,
-    abi: ERC20_ABI,
-    functionName: 'balanceOf',
-    args: [address],
-  })) : [];
+  const balanceReadContracts = address?.startsWith('0x')
+    ? tokens.map(token => ({
+        address: token.address,
+        abi: ERC20_ABI,
+        functionName: 'balanceOf',
+        args: [address],
+      }))
+    : [];
 
   const fetchEtherBalance = async () => {
     if (!address?.startsWith('0x')) return;
 
     const balance = await publicClient.getBalance({ address });
 
     return balance;
-  }
+  };
 
   const fetchBalances = async () => {
-      const balanceData = await publicClient.multicall({ contracts: balanceReadContracts });
-      const etherBalance = await fetchEtherBalance();
-  
-      const balanceMap: BalanceMap = new Map();
-    
-      for(let i=0; i<balanceReadContracts.length; i++) {
-        try {
-          const { status: balanceStatus, ...balanceArgs } = balanceData[i];
-  
-          const { address, decimals } = tokens[i];
-          const tokenAddress = address.toLowerCase() as `0x${string}`;
-          const isNativeToken = tokenAddress === ETH_CONTRACT_ADDRESS.toLowerCase();
-
-          if (isNativeToken && etherBalance) {
-            balanceMap.set(tokenAddress, formatTokenAmount(etherBalance.toString(), 18));
-            continue;
-          }
-
-          if (balanceStatus === 'success') {
-            const { result: rawBalance } = balanceArgs as { result: bigint };
-
-            const formattedTokenBalance = formatTokenAmount(rawBalance.toString(), decimals);
-      
-            balanceMap.set(tokenAddress, formattedTokenBalance);
-            continue;
-          }
-
-          balanceMap.set(tokenAddress, '0');
-                  
-        } catch (error) {
-          console.error(error);
+    const usdPricesForTokens = await Promise.allSettled(
+      tokens.map(token => sifi.getUsdPrice(chainId, token.address))
```

Comment:
Could we do this after we have fetched multicall token balances, so we only call this endpoint for tokens with wallet balance? Would save a lot of API calls.

Reply:
Alternatively, we could do it the other way round, where we do the multicall after fetching the USD balances?
We are still getting easily rate limited on that so worth experimenting with.